lib/spack/spack/relocate.py

```diff
@@ -0,0 +1,317 @@
+    id = None
+    nrpaths = []
+    ndeps = []
+    if idpath:
+        id = '@rpath/%s' % os.path.basename(idpath)
+    for rpath in rpaths:
+        if re.match(old_dir, rpath):
+            rel = os.path.relpath(rpath, start=os.path.dirname(path_name))
+            nrpaths.append('@loader_path/%s' % rel)
+        else:
+            nrpaths.append(rpath)
+    for dep in deps:
+        if re.match(old_dir, dep):
+            rel = os.path.relpath(dep, start=os.path.dirname(path_name))
+            ndeps.append('@loader_path/%s' % rel)
+        else:
+            ndeps.append(dep)
+    return nrpaths, ndeps, id
+
+
+def macho_replace_paths(old_dir, new_dir, rpaths, deps, idpath):
+    id = None
+    nrpaths = []
+    ndeps = []
+    if idpath:
+        id = idpath.replace(old_dir, new_dir)
+    for rpath in rpaths:
+        nrpath = rpath.replace(old_dir, new_dir)
+        nrpaths.append(nrpath)
+    for dep in deps:
+        ndep = dep.replace(old_dir, new_dir)
+        ndeps.append(ndep)
+    return nrpaths, ndeps, id
+
+
+def modify_macho_object(path_name, old_dir, new_dir, relative):
+    """
+    Modify MachO binaries by changing rpaths,and id and dependency lib paths.
+    Examines the output of otool -l for these three patterns
+    cmd LC_ID_DYLIB
+    cmdsize 160
+    name /Users/gartung/spack-macdev/opt/spack/darwin-x86_64/clang-7.0.2
+    -apple/tcl-8.6.5-xfeydlhaojmei6iws2rnxndvriym242k/lib/libtcl8.6.dylib
+    (offset 24)
+    cmd LC_LOAD_DYLIB
+    cmdsize 160
+    name /Users/gartung/spack-macdev/opt/spack/darwin-x86_64/clang-7.0.2
+    -apple/zlib-1.2.8-cyvcqvrzlgurne424y55hxvfucvz2354/lib/libz.1.dylib
+    (offset 24)
+    cmd LC_RPATH
+    cmdsize 128
+    path /Users/gartung/spack-macdev/opt/spack/darwin-x86_64/clang-7.0.2
+    -apple/xz-5.2.2-d4ecxpuzf2g3ycz3cnj3xmdj7zdnuqwb/lib
+    (offset 12)
+    the old install dir in LC_LOAD_DYLIB is replaced with the new install dir
+    using install_name_tool -id newid binary
+    the old install dir in LC_LOAD_DYLIB is replaced with the new install dir
+    using install_name_tool -change old new binary
+    the old install dir in LC_RPATH is replaced with the new install dir using
+    install_name_tool  -rpath old new binary
+    """
+    if 'libgcc_' in path_name:
+        return
+    rpaths, deps, idpath = macho_get_paths(path_name)
+    id = None
+    nrpaths = []
+    ndeps = []
+    if relative:
+        nrpaths, ndeps, id = macho_make_paths_rel(path_name,
+                                                  old_dir, rpaths,
+                                                  deps, idpath)
+    else:
+        nrpaths, ndeps, id = macho_replace_paths(old_dir, new_dir, rpaths,
+                                                 deps, idpath)
+    st = os.stat(path_name)
+    wmode = os.access(path_name, os.W_OK)
+    if not wmode:
+        try:
+            os.chmod(path_name, st.st_mode | stat.S_IWUSR)
+        except OSError as e:
+            tty.warn(e)
```

Comment:
Sorry I didn't mean to comment on the approach of using the system files here other than to offer a suggestion if that was the reason you are printing warnings here. My main point in this case is that you shouldn't need to check for write permissions or print warnings if they aren't present. Spack should just fail if the permissions aren't set properly and report that; I think that would be achieved by not catching the ```ProcessError``` exceptions below.